### ISSUE TITLE
Speed comparison of mkl-dnn vs eigenTensor

### ISSUE THREAD

**Issue by WilliamTambellini** [intel/mkl-dnn#516]
Hello, 
This is just a call for review in order to validate a minimalist benchmark of softmax fwd between mkl-dnn and 'eigenTensor' (the default engine of TensorFlow).

On a simple avx2 CPU I got : 
ETensor L=0 softmaxAxis=0          a=30000          b=8        5489 usecs/fwd
ETensor L=1 softmaxAxis=0          a=30000          b=8        5763 usecs/fwd
MKL-DNN     softmaxAxis=0          a=30000          b=8       19358 usecs/fwd
ETensor L=0 softmaxAxis=1          a=8          b=30000        5842 usecs/fwd
ETensor L=1 softmaxAxis=1          a=8          b=30000        5405 usecs/fwd
MKL-DNN     softmaxAxis=1          a=8          b=30000         207 usecs/fwd

That could perhaps confirm : 
https://intel.github.io/mkl-dnn/dev_guide_softmax.html
"
 Optimized: 2D case, tensor A×B, softmax axis 1 (B), format tag mkldnn_ab
 Non-optimized: 2D case, tensor A×B, softmax axis 0 (A), format tag mkldnn_ab, and B≠1
 Non-optimized: 2D case, tensor A×B, softmax axis 1 (B), format tag mkldnn_ba, and A≠1
"

The benchmark (for input 8x30k and softmax on B) shows a 20x speed up over EigenTensor : that part needs confirmation/reviewing. 
I will check with the eigen team for any bug/optimization in my code.
Could anyone here just check the mkldnn side of this : 
https://gist.github.com/WilliamTambellini/8294f211800e16791d47f3cf59472a49
?
Kind
W.
 

-----------------------------------------------------------------------------

### Environment
* CPU make and model : Intel(R) Core(TM) i7-4600U CPU @ 2.10GHz
* OS version (`uname -a`) : Linux PARLT42X4Q12 3.10.0-862.11.6.el7.x86_64 #1 SMP Tue Aug 14 21:49:04 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
* Compiler version (`gcc --version`) 6.1.0
* CMake version (`cmake --version`) : 3.13
* CMake output log
-- The C compiler identification is GNU 6.1.0
-- The CXX compiler identification is GNU 6.1.0
-- Check for working C compiler: gcc-6.1.0/bin/gcc
-- Check for working C compiler: gcc-6.1.0/bin/gcc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: gcc-6.1.0/bin/g++
-- Check for working CXX compiler: gcc-6.1.0/bin/g++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Found OpenMP_C: -fopenmp (found version "4.5") 
-- Found OpenMP_CXX: -fopenmp (found version "4.5") 
-- Found OpenMP: TRUE (found version "4.5")  
-- GPU support is disabled
 VTuneel Amplifier JIT profiling disabled

* git hash (`git log -1 --format=%H`)
038c9c7e58311b414452b89b3f3bbc9566e8b7df

### Steps to reproduce
Build and run the provided gist example : 
g++ -std=c++11 -mavx2 -mfma -march=native 
 -DEIGEN_NO_DEBUG -DNDEBUG -fopenmp -O3 
 -I eigen-bc7e634886a4 
 -I $mkldir/include -L $mkldir/lib64 -l mkldnn 
 eigen_vs_mkldnn.cpp -o eigen_vs_mkldnn

### Actual behavior
20 x speed up

### Expected behavior
some speed up ?

**Comment by emfomenk** [comment]
Hi @WilliamTambellini,

I will take a look on the benchmark tomorrow (which I think is fine, at least from the perspective of A case being slower than B exactly due to the reason you pointed), but could you please elaborate on how useful the case of softmax along A-axis is?
When we optimized the softmax, we intentionally focused on the case where softmax-axis is the innermost one -- we've never seen a topology that uses something other than that.

**Comment by rsdubtso** [comment]
The MKL-DNN part of the code seems fine to me.

**Comment by vpirogov** [comment]
Intel MKL-DNN has validation and benchmarking tool called [benchdnn](https://github.com/intel/mkl-dnn/tree/master/tests/benchdnn) that we typically use to validate the performance.

On 2 cores of Intel Xeon E5-2699 v4 I see the numbers in the same ballpark for both benchmarks. 
``` 
$ MKLDNN_VERBOSE=1 ./a.out 
mkldnn_verbose,info,Intel(R) MKL-DNN v1.0.0 (Git Hash c0457f7d825f1fea5cd82ae68998dc6dd56216ab),Intel(R) Advanced Vector Extensions 2 (Intel(R) AVX2)
mkldnn_verbose,exec,cpu,softmax,ref:any,forward_inference,data_f32::blocked:ab:f0 diff_undef::undef::f0,axis:0,30000x8,31.0601
MKL-DNN
softmaxAxis=0          a=30000          b=8
mkldnn_verbose,exec,cpu,softmax,ref:any,forward_inference,data_f32::blocked:ab:f0 diff_undef::undef::f0,axis:0,30000x8,30.5791
       30605 usecs/fwd
mkldnn_verbose,exec,cpu,softmax,jit:avx2,forward_inference,data_f32::blocked:ab:f0 diff_undef::undef::f0,axis:1,8x30000,0.698975
MKL-DNN
softmaxAxis=1          a=8          b=30000mkldnn_verbose,exec,cpu,softmax,jit:avx2,forward_inference,data_f32::blocked:ab:f0 diff_undef::undef::f0,axis:1,8x30000,0.39209
         414 usecs/fwd
```

```
$ ./tests/benchdnn/benchdnn --softmax --mode=P --tag=ab 30000x8     
Output template: perf,%engine%,%desc%,%-time%,%0time%
perf,cpu,--softmax --tag=ab 30000x8,1.91455,1.93196
tests:1 passed:0 skipped:0 mistrusted:0 unimplemented:0 failed:0
total perf: min(ms):1.91455 avg(ms):1.93196
$ ./tests/benchdnn/benchdnn --softmax --mode=P --tag=ab 8x30000     
Output template: perf,%engine%,%desc%,%-time%,%0time%
perf,cpu,--softmax --tag=ab 8x30000,0.179688,0.181241
tests:1 passed:0 skipped:0 mistrusted:0 unimplemented:0 failed:0
total perf: min(ms):0.179688 avg(ms):0.181241
```

**Comment by emfomenk** [comment]
Please note that benchdnn by default always assumes `softmax_axis=1`, so the run that matches the example should be:
``` bash
$ export OMP_NUM_THREADS=2 KMP_AFFINITY=granularity=fine,compact
$ ./benchdnn --softmax --mode=p --tag=ab --axis=0 30000x8 --axis=1 8x30000
Output template: perf,%engine%,%desc%,%-time%,%0time%
perf,cpu,--softmax --tag=ab --axis=0 30000x8,15.582,15.615
perf,cpu,--softmax --tag=ab 8x30000,0.540039,0.545098
tests:2 passed:0 skipped:0 mistrusted:0 unimplemented:0 failed:0
total perf: min(ms):16.1221 avg(ms):16.1601
```

**Comment by vpirogov** [comment]
Indeed I missed the axis and misread the first number when converting from us to ms. Here's what I see with GCC:
```
$ OMP_NUM_THREADS=2 ./tests/benchdnn/benchdnn --softmax --mode=P --t
ag=ab --axis=0 30000x8 --axis=1 8x30000                                      
Output template: perf,%engine%,%desc%,%-time%,%0time%
perf,cpu,--softmax --tag=ab --axis=0 30000x8,23.0012,23.1355
perf,cpu,--softmax --tag=ab 8x30000,0.182373,0.184103
tests:2 passed:0 skipped:0 mistrusted:0 unimplemented:0 failed:0
total perf: min(ms):23.1836 avg(ms):23.3196
```

**Comment by WilliamTambellini** [comment]
Ok thank you everybody.
I made a mistake: I forgot to set OMP_NUM_THREADS to do a fair comparison: both should use the same number of threads I suppose.
By imposing 1 OMP thread, the speedup seems to be "only" 10x:
..
omp_get_max_threads: 1
...
ETensor L=0 softmaxAxis=1          a=8          b=30000        5679 usecs/fwd
ETensor L=1 softmaxAxis=1          a=8          b=30000        5279 usecs/fwd
MKL-DNN softmaxAxis=1              a=8          b=30000         537 usecs/fwd

The ticket in eigen bugtracker : 
http://eigen.tuxfamily.org/bz/show_bug.cgi?id=1733

If there is not much to add to the mkldnn side then I guess this github issue could be closed.

**Comment by vpirogov** [comment]
Considering that Eigen Tensor does not provide a softmax function it should probably be a defect against Tensorflow, which was used as a reference implementation for softmax.

**Comment by vpirogov** [comment]
@WilliamTambellini, I noticed that Eigen performance has improved according to Rasmus. Do you have updated datapoint to show where we stand now?

**Comment by WilliamTambellini** [comment]
Hello @vpirogov  
Voila, I ve just rerun my benchmark on the very same CPU using the latest eigen master (8071cda5714d), OMP_NUM_THREADS=1, and indeed:
ETensor L=0   softmaxAxis=0          a=30000          b=8         933
ETensor L=1   softmaxAxis=0          a=30000          b=8        1375
MKL-DNN       softmaxAxis=0          a=30000          b=8       20420
ETensor L=0   softmaxAxis=1          a=8          b=30000        1374
ETensor L=1   softmaxAxis=1          a=8          b=30000         922
MKL-DNN       softmaxAxis=1          a=8          b=30000         528
which is similar to the speedups reported by Rasmus. 
Tks
W.

**Comment by vpirogov** [comment]
Thank you very much. So we are within 2-3x now.
